File: App/src/main/java/com/drownedinsound/ui/home/HomeScreenPresenter.java
package com.drownedinsound.ui.home;

import com.drownedinsound.data.DisBoardRepo;
import com.drownedinsound.data.generatered.BoardPostList;
import com.drownedinsound.qualifiers.ForIoScheduler;
import com.drownedinsound.qualifiers.ForMainThreadScheduler;
import com.drownedinsound.ui.home.postList.BoardPostListContract;

import java.util.HashMap;
import java.util.List;
import java.util.Map;

import rx.Observer;
import rx.Scheduler;

public class HomeScreenPresenter implements HomeScreenContract.Presenter {

    private final HomeScreenContract.View homeScreenView;

    private final DisBoardRepo disBoardRepo;

    private final Scheduler mainThreadScheduler;

    private final Scheduler backgroundThreadScheduler;

    private final Map<String, BoardPostListContract.Presenter> boardPostListPresenterMap
            = new HashMap<>();

    public HomeScreenPresenter(
            HomeScreenContract.View homeScreenView,
            DisBoardRepo disBoardRepo,
            @ForMainThreadScheduler Scheduler mainThreadScheduler,
            @ForIoScheduler Scheduler backgroundThreadScheduler) {
        this.homeScreenView = homeScreenView;
        this.disBoardRepo = disBoardRepo;
        this.mainThreadScheduler = mainThreadScheduler;
        this.backgroundThreadScheduler = backgroundThreadScheduler;
    }

    @Override
    public void onViewCreated() {
        getAllBoardPostLists();
    }

    private void getAllBoardPostLists() {
        disBoardRepo.getAllBoardPostLists()
                .subscribeOn(backgroundThreadScheduler)
                .observeOn(mainThreadScheduler)
                .subscribe(new Observer<List<BoardPostList>>() {
                    @Override
                    public void onCompleted() {

                    }

                    @Override
                    public void onError(Throwable e) {

                    }

                    @Override
                    public void onNext(List<BoardPostList> boardPostLists) {
                        homeScreenView.showBoardPostLists(boardPostLists);
                    }
                });
    }

    @Override
    public void addBoardListPresenter(String type, BoardPostListContract.Presenter presenter) {
        boardPostListPresenterMap.put(type, presenter);
        presenter.onViewCreated();
    }

    @Override
    public void removeBoardPostListView(String type) {
        BoardPostListContract.Presenter removedListPresenter = boardPostListPresenterMap
                .remove(type);
        removedListPresenter.onViewDestroyed();
    }

    @Override
    public void handleListDisplayed(String type) {
        BoardPostListContract.Presenter boardPostListPresenter = boardPostListPresenterMap
                .get(type);
        boardPostListPresenter.onViewDisplayed();
    }

    @Override
    public void handlePageTabReselected(int position) {

    }
}

File: App/src/main/kotlin/com/drownedinsound/ui/home/HomeScreenPresenter.kt
package com.drownedinsound.ui.home

import com.drownedinsound.core.DisBoardsAppModule
import com.drownedinsound.data.DisBoardRepo
import com.drownedinsound.ui.home.postList.BoardPostListContract
import rx.Scheduler
import java.util.*
import javax.inject.Inject
import javax.inject.Named

class HomeScreenPresenter @Inject constructor(
        private val homeScreenView: HomeScreenContract.View,
        private val disBoardRepo: DisBoardRepo,
        @param:Named(DisBoardsAppModule.MAIN_THREAD_SCHEDULER) private val mainThreadScheduler: Scheduler,
        @param:Named(DisBoardsAppModule.BACKGROUND_THREAD_SCHEDULER) private val backgroundThreadScheduler: Scheduler) : HomeScreenContract.Presenter {

    private val listPresenterMap = HashMap<String, BoardPostListContract.Presenter>()

    override fun onViewCreated() {
        getAllBoardPostLists()
    }

    private fun getAllBoardPostLists() {
        disBoardRepo.allBoardPostLists
                .subscribeOn(backgroundThreadScheduler)
                .observeOn(mainThreadScheduler)
                .subscribe(homeScreenView::showBoardPostLists)
    }

    override fun addBoardListPresenter(type: String, presenter: BoardPostListContract.Presenter) {
        listPresenterMap.put(type, presenter)
        presenter.onViewCreated()
    }

    override fun removeBoardPostListView(type: String) {
        listPresenterMap.remove(type)?.onViewDestroyed()
    }

    override fun handleListDisplayed(type: String) {
        listPresenterMap[type]?.onViewDisplayed()
    }

    override fun handlePageTabReselected(position: Int) {

    }
}